classes/local/step/connector_sftp.php

```diff
@@ -0,0 +1,302 @@
+
+        // If no key file is given, then a password is required.
+        if ($noprivfile && empty($config->password)) {
+            $errors['config_password'] = get_string(
+                'config_field_missing',
+                'tool_dataflows',
+                get_string('password'),
+                true
+            );
+        }
+        if (empty($config->source)) {
+            $errors['config_source'] = get_string(
+                'config_field_missing',
+                'tool_dataflows',
+                get_string('connector_sftp:source', 'tool_dataflows'),
+                true
+            );
+        }
+        if (empty($config->target)) {
+            $errors['config_target'] = get_string(
+                'config_field_missing',
+                'tool_dataflows',
+                get_string('connector_sftp:target', 'tool_dataflows'),
+                true
+            );
+        }
+
+        $hasremote = true;
+        // Check that at least one file config has an sftp:// scheme.
+        if (!empty($config->source) && !empty($config->target)) {
+            // Check if the source or target is an expression, and evaluate it if required.
+            $sourceremote = helper::path_has_scheme($config->source, self::SFTP_PREFIX);
+            $targetremote = helper::path_has_scheme($config->target, self::SFTP_PREFIX);
+            $hasremote = $sourceremote || $targetremote;
+        }
+        if (!$hasremote) {
+            $errormsg = get_string('connector_sftp:missing_remote', 'tool_dataflows', null, true);
+            $errors['config_source'] = $errors['config_source'] ?? $errormsg;
+            $errors['config_target'] = $errors['config_target'] ?? $errormsg;
+        }
+
+        return empty($errors) ? true : $errors;
+    }
+
+    /**
+     * Perform any extra validation that is required only for runs.
+     *
+     * @return true|array Will return true or an array of errors.
+     */
+    public function validate_for_run() {
+        $config = $this->get_config();
+
+        $errors = [];
+
+        $error = helper::path_validate($config->source);
+        if ($error !== true) {
+            $errors['config_source'] = $error;
+        }
+
+        $error = helper::path_validate($config->target);
+        if ($error !== true) {
+            $errors['config_target'] = $error;
+        }
+
+        if (!empty($config->privkeyfile)) {
+            $error = helper::path_validate($config->privkeyfile);
+            if ($error !== true) {
+                $errors['config_privkeyfile'] = $error;
+            }
+        }
+
+        return $errors ?: true;
+    }
+
+    /**
+     * Executes the step
+     *
+     * Performs an SFTP call according to config parameters.
+     *
+     * @param  mixed|null $input
+     * @return mixed
+     */
+    public function execute($input = null) {
+        $config = $this->get_config();
+
+        $this->log("Connecting to $config->host:$config->port");
```

Comment:
```suggestion
        $this->log("Connecting to {$config->host}:{$config->port}");
```

I'm testing out an alternative way to weave it into the main string. What do you think? Would this suggestion make it easier to identify the individual variables?